example/src/main/java/io/netty/example/http2/tiles/http2/Http2RequestHandler.java

```diff
@@ -0,0 +1,138 @@
+
+        QueryStringDecoder queryString = new QueryStringDecoder(request.uri());
+        String streamId = streamId(request);
+
+        // Check arguments: path must match and latency parameter must be
+        // present
+        if (!PATH.equals(queryString.path()) || missing(queryString, LATENCY)) {
+            sendDummy(ctx, streamId);
+            return;
+        }
+
+        int latency = latency(queryString);
+
+        if (missing(queryString, X) && missing(queryString, Y)) {
+            handlePage(ctx, streamId, latency);
+        } else {
+            handleImage(queryString, ctx, streamId, latency);
+        }
+    }
+
+    private void handleImage(QueryStringDecoder query, ChannelHandlerContext ctx, String streamId, int latency) {
+        int x = toInt(query.parameters().get(X).get(0), 0);
+        int y = toInt(query.parameters().get(Y).get(0), 0);
+        byte[] image = ImageCache.instance().image(x, y);
+        FullHttpResponse response = new DefaultFullHttpResponse(HTTP_1_1, OK, wrappedBuffer(image));
+        response.headers().set(CONTENT_TYPE, "image/jpeg");
+        sendResponse(ctx, streamId, latency, response);
+    }
+
+    private void handlePage(ChannelHandlerContext ctx, String streamId, int latency) {
+        ByteBuf content = Unpooled.wrappedBuffer(Html.HEADER, Html.body(latency), Html.FOOTER);
+        FullHttpResponse response = new DefaultFullHttpResponse(HTTP_1_1, OK, content);
+        response.headers().set(CONTENT_TYPE, "text/html; charset=UTF-8");
+        sendResponse(ctx, streamId, latency, response);
+    }
+
+    protected void sendResponse(final ChannelHandlerContext ctx, String streamId, int latency,
+            final FullHttpResponse response) {
+        setContentLength(response, response.content().readableBytes());
+        setStreamId(response, streamId);
+        ctx.executor().schedule(new Runnable() {
+            @Override
+            public void run() {
+                ctx.writeAndFlush(response);
+            }
+        }, latency, TimeUnit.MILLISECONDS);
+    }
+
+    private int latency(QueryStringDecoder queryString) {
+        int latency = toInt(queryString.parameters().get(LATENCY).get(0), 0);
+        // Make sure that latency is not too big
+        return abs(latency > 1000 ? 0 : latency);
+    }
+
+    private String streamId(FullHttpRequest request) {
+        return request.headers().get(HttpUtil.ExtensionHeaderNames.STREAM_ID.text());
+    }
+
+    private void setStreamId(FullHttpResponse response, String streamId) {
+        response.headers().set(HttpUtil.ExtensionHeaderNames.STREAM_ID.text(), streamId);
+    }
+
+    private boolean missing(QueryStringDecoder query, String string) {
+        List<String> values = query.parameters().get(string);
+        return query.parameters() == null || values == null || values.size() == 0 || EMPTY_STRING.equals(values.get(0));
```

Comment:
You used `query.parameters()` on the line before. It can't be `null` here.

Is there an iterator method...or 1 that returns the first value so we don't have to deal with a list?

Also consider changing this to `getQueryParameter(...)` (or something similiar) and just return null if it doesn't exist.  This would help avoid duplication is building a collection to see if it exists...and then after that building the collection again to get the first element.